### ISSUE TITLE
Render and promise resolution Issues in development

### ISSUE THREAD

**Issue by alevar** [GMOD/jbrowse-react-linear-genome-view-vite-demo#2]
Hi, thank you for building the JBrowse, supporting it and providing demos! I was able to clone and run the app and reproduce the same view as in your online version. However, while in development mode (npm/yarn run dev), I get the following error

```
Warning: Cannot update a component (`View`) while rendering a different component (`Unknown`). To locate the bad setState() call inside `Unknown`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render Component Stack: 
    observerComponent MobX
    div unknown:0
    observerComponent MobX
    div unknown:0
    node_modules chunk-CLDML6YG.js:1700
    Paper2 chunk-CLDML6YG.js:14944
    observerComponent MobX
    div unknown:0
    node_modules chunk-CLDML6YG.js:1700
    ScopedCssBaseline2 chunk-CLDML6YG.js:50793
    div unknown:0
    observerComponent MobX
    DefaultPropsProvider chunk-CLDML6YG.js:8064
    RtlProvider chunk-CLDML6YG.js:8035
    ThemeProvider2 chunk-CLDML6YG.js:7969
    ThemeProvider3 chunk-CLDML6YG.js:8167
    ThemeProviderNoVars chunk-CLDML6YG.js:10273
    ThemeProvider4 chunk-CLDML6YG.js:11893
    observerComponent MobX
    View App.tsx:16
```
    
 Despite the error, the browser loads along with all tracks. However, on hover over elements of alignment or gene tracks, the tracks disappear with the following error:
 
 ```
Uncaught Error: Element type is invalid. Received a promise that resolves to: [object Object]. Lazy element type must resolve to a class or function.
    React 10
    workLoop scheduler.development.js:266
    flushWork scheduler.development.js:239
    performWorkUntilDeadline scheduler.development.js:533

The above error occurred in one of your React components:

Suspense
observerComponent@http://localhost:8000/node_modules/.vite/deps/chunk-J2UKEEFY.js?v=6c70a36a:455:12
Suspense
div
observerComponent@http://localhost:8000/node_modules/.vite/deps/chunk-J2UKEEFY.js?v=6c70a36a:455:12
div
div
observerComponent@http://localhost:8000/node_modules/.vite/deps/chunk-J2UKEEFY.js?v=6c70a36a:455:12
Suspense
div
div
observerComponent@http://localhost:8000/node_modules/.vite/deps/chunk-J2UKEEFY.js?v=6c70a36a:455:12
ErrorBoundary@http://localhost:8000/node_modules/.vite/deps/chunk-PIK6533K.js?v=6c70a36a:2137:9
div
node_modules/@emotion/react/dist/emotion-element-7a1343fa.browser.development.esm.js/withEmotionCache2/<@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:1700:50
Paper2@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:14944:37
observerComponent@http://localhost:8000/node_modules/.vite/deps/chunk-J2UKEEFY.js?v=6c70a36a:455:12
div
observerComponent@http://localhost:8000/node_modules/.vite/deps/chunk-J2UKEEFY.js?v=6c70a36a:455:12
div
observerComponent@http://localhost:8000/node_modules/.vite/deps/chunk-J2UKEEFY.js?v=6c70a36a:455:12
Suspense
div
node_modules/@emotion/react/dist/emotion-element-7a1343fa.browser.development.esm.js/withEmotionCache2/<@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:1700:50
Paper2@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:14944:37
div
node_modules/@emotion/react/dist/emotion-element-7a1343fa.browser.development.esm.js/withEmotionCache2/<@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:1700:50
Paper2@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:14944:37
observerComponent@http://localhost:8000/node_modules/.vite/deps/chunk-J2UKEEFY.js?v=6c70a36a:455:12
div
node_modules/@emotion/react/dist/emotion-element-7a1343fa.browser.development.esm.js/withEmotionCache2/<@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:1700:50
ScopedCssBaseline2@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:50793:37
div
observerComponent@http://localhost:8000/node_modules/.vite/deps/chunk-J2UKEEFY.js?v=6c70a36a:455:12
DefaultPropsProvider@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:8064:30
RtlProvider@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:8035:21
ThemeProvider2@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:7969:7
ThemeProvider3@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:8167:7
ThemeProviderNoVars@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:10273:29
ThemeProvider4@http://localhost:8000/node_modules/.vite/deps/chunk-CLDML6YG.js?v=6c70a36a:11893:24
observerComponent@http://localhost:8000/node_modules/.vite/deps/chunk-J2UKEEFY.js?v=6c70a36a:455:12
View@http://localhost:8000/src/App.tsx:30:37
```

React will try to recreate this component tree from scratch using the error boundary you provided, ErrorBoundary.

None of these errors are present, however, when in production mode with npm/yarn run build && npm yarn run preview. Is this behavior expected for the development mode, and if not, how may you advise mitigating this issue?

Thank you,

Ales

**Comment by cmdcolin** [comment]
thanks for reporting this, i can reproduce it. I'll see if it can find a fix

**Comment by cmdcolin** [comment]
this is the second time we've had bugs in dev mode vs prod mode lately....vite is a funny system where dev mode uses a totally different bundler (esbuild) vs production (rollup) so it's weird sometimes

**Comment by alevar** [comment]
Thank you for looking into it and confirming the bug being due to dev vs prod!

I wanted to post a small update though. I eventually got jbrowse to work with my data - pulling the genome and gtf files (the only I need for my use) from the backend api. The genome is small (viral 10kb with a few dozens alternatively spliced transcripts in the entire genome). The issue of failing Feature track in dev mode is no longer present on this data. There is a number of changes however:
- IndexedFastaAdapter instead of gzipped
- tiny genome size
- GFF was standardized from GTF using gffread, piped into `sort` prior to compressing with bgzip and indexing with tabix

Happy to provide any additional information that may aid debugging and will update here if I find out what made the issue go away for me.

Thank you,

**Comment by cmdcolin** [comment]
glad you got it working. I think I have the repro needed for this particular case

if you run into any other issues let me know. that sounds like a good setup as far as i can tell!

in some cases, you can load the data entirely into what we call "FromConfigAdapters" so that you can provide smallish datasets directly into the 'config', or you can make custom data adapters to fetch data directly from a REST API (example for writing those https://jbrowse.org/jb2/docs/developer_guides/creating_adapter/)

**Comment by cmdcolin** [comment]
should be fixed in v2.16.0

note that the first thing you mentioned also  ("Cannot update a component (`View`) while rendering a different component (`Unknown`)") is a known issue that is just related to that onPatch callback calling setState, react doesn't seem to like that,  but it is totally optional for the demo to use that. generally we suggest leveraging the mobx "observer"  concept to get the state rather than something like onpatch. we will change it if we find a better way to do things that fixes this warning.

**Comment by cmdcolin** [comment]
lemme know if you run into anything else or didn't get a fix for this! should be able to get the fix by recloning demo or  editing package.json to get the new version or whatnot
